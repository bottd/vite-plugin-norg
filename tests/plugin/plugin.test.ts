import { vi } from 'vitest';
import { fileURLToPath } from 'node:url';
import { join, dirname } from 'node:path';
import { norgPlugin } from '../../src/plugin/index.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
describe('vite-plugin-norg', () => {
  describe('HTML', () => {
    const htmlPlugin = norgPlugin({ mode: 'html', include: ['**/*.norg'] });

    it('should generate correct output for basic norg file', async () => {
      const fixturePath = join(__dirname, '../fixtures/basic.norg');
      const result = await htmlPlugin.load(fixturePath);
      expect(result).toMatchSnapshot();
    });

    it('should generate correct output for code blocks', async () => {
      const fixturePath = join(__dirname, '../fixtures/code-blocks.norg');
      const result = await htmlPlugin.load(fixturePath);
      expect(result).toMatchSnapshot();
    });

    it('should generate correct output for all heading levels', async () => {
      const fixturePath = join(__dirname, '../fixtures/headings.norg');
      const result = await htmlPlugin.load(fixturePath);
      expect(result).toMatchSnapshot();
    });

    it('should generate correct output for image syntax', async () => {
      const fixturePath = join(__dirname, '../fixtures/images.norg');
      const result = await htmlPlugin.load(fixturePath);
      expect(result).toMatchSnapshot();
    });

    it('should return null for non-norg files', async () => {
      const result = await htmlPlugin.load('console.log("test");', 'test.js');
      expect(result).toBeUndefined();
    });

    it('should have correct plugin name', () => {
      expect(htmlPlugin.name).toBe('vite-plugin-norg');
    });

    it('should enforce pre', () => {
      expect(htmlPlugin.enforce).toBe('pre');
    });

    it('should resolve .norg files', async () => {
      // Mock the plugin context
      const mockContext = {
        resolve: vi.fn().mockResolvedValue({ id: 'test.norg' }),
      };

      expect(await htmlPlugin.resolveId.call(mockContext, 'test.norg')).toBe('test.norg');
      expect(await htmlPlugin.resolveId.call(mockContext, 'test.js')).toBeNull();
    });
  });

  describe('Svelte Mode', () => {
    const sveltePlugin = norgPlugin({ mode: 'svelte', include: ['**/*.norg'] });

    it('should generate correct component for basic norg file', async () => {
      const fixturePath = join(__dirname, '../fixtures/basic.norg');
      const result = await sveltePlugin.load(fixturePath);
      expect(result).toMatchSnapshot();
    });

    it('should generate correct component for code blocks', async () => {
      const fixturePath = join(__dirname, '../fixtures/code-blocks.norg');
      const result = await sveltePlugin.load(fixturePath);
      expect(result).toMatchSnapshot();
    });

    it('should generate correct component for all heading levels', async () => {
      const fixturePath = join(__dirname, '../fixtures/headings.norg');
      const result = await sveltePlugin.load(fixturePath);
      expect(result).toMatchSnapshot();
    });
  });

  describe('React Mode', () => {
    const reactPlugin = norgPlugin({ mode: 'react', include: ['**/*.norg'] });

    it('should generate correct component for basic norg file', async () => {
      const fixturePath = join(__dirname, '../fixtures/basic.norg');
      const result = await reactPlugin.load(fixturePath);
      expect(result).toMatchSnapshot();
    });

    it('should generate correct component for code blocks', async () => {
      const fixturePath = join(__dirname, '../fixtures/code-blocks.norg');
      const result = await reactPlugin.load(fixturePath);
      expect(result).toMatchSnapshot();
    });

    it('should generate correct component for all heading levels', async () => {
      const fixturePath = join(__dirname, '../fixtures/headings.norg');
      const result = await reactPlugin.load(fixturePath);
      expect(result).toMatchSnapshot();
    });
  });
});
