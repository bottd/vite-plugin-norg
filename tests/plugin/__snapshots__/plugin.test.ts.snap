// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vite-plugin-norg > HTML > should generate correct output for all heading levels 1`] = `
"export const metadata = {"title":"\\"Headings\\""};
export const html = "<h1 id=\\"h1\\">H1</h1>\\n<p>Content 1</p><h2 id=\\"h2\\">H2</h2>\\n<p>Content 2</p><h3 id=\\"h3\\">H3</h3>\\n<p>Content 3</p><h4 id=\\"h4\\">H4</h4>\\n<p>Content 4</p><h5 id=\\"h5\\">H5</h5>\\n<p>Content 5</p><h6 id=\\"h6\\">H6</h6>\\n<p>Content 6</p>";
export default { metadata, html };"
`;

exports[`vite-plugin-norg > HTML > should generate correct output for basic norg file 1`] = `
"export const metadata = {"author":"Drake Bott","tags":["\\"test\\", \\"basic\\""],"title":"Basic Norg"};
export const html = "<h1 id=\\"main-title\\">Main Title</h1>\\n<p>Text with <i>bold</i>, <em>italic</em>, <code>code</code>, and <s>strikethrough</s>.</p><h2 id=\\"lists\\">Lists</h2>\\n<ul><li>Item 1</li></ul><ul><li>Item 2</li></ul><ol><li>Number 1</li></ol><ol><li>Number 2</li></ol><h2 id=\\"link\\">Link</h2>\\n<p><a href=\\"https://example.com\\" target=\\"_blank\\">link text</a></p>";
export default { metadata, html };"
`;

exports[`vite-plugin-norg > HTML > should generate correct output for code blocks 1`] = `
"export const metadata = {"title":"\\"Code Test\\""};
export const html = "<h1 id=\\"code-blocks\\">Code Blocks</h1>\\n<h2 id=\\"javascript\\">JavaScript</h2>\\n<pre class=\\"language-javascript\\"><code class=\\"language-javascript\\">function hello() { return &quot;world&quot;; }\\n</code></pre><h2 id=\\"python\\">Python </h2>\\n<pre class=\\"language-python\\"><code class=\\"language-python\\">def add(a, b): return a + b\\n</code></pre><h2 id=\\"plain\\">Plain</h2>\\n<pre><code>plain text\\n</code></pre>";
export default { metadata, html };"
`;

exports[`vite-plugin-norg > HTML > should generate correct output for image syntax 1`] = `
"export const metadata = {};
export const html = "<h1 id=\\"images\\">Images</h1>\\n<p>Here are some images:</p><p><a href=\\"https://example.com/image1.png\\" target=\\"_blank\\">Image Alt Text</a></p><p><a href=\\"/local/path/image.jpg\\">Local Image</a></p><p><a href=\\"./relative/image.png\\">./relative/image.png</a>[]</p>";
export default { metadata, html };"
`;

exports[`vite-plugin-norg > HTML > should generate correct output for minimal norg files 1`] = `
"export const metadata = {};
export const html = "<h1 id=\\"simple-title\\">Simple Title</h1>\\n<p>Basic content.</p>";
export default { metadata, html };"
`;

exports[`vite-plugin-norg > React Mode > should generate correct component for all heading levels 1`] = `
"import React from "react";
export const metadata = {"title":"\\"Headings\\""};
const htmlContent = "<h1 id=\\"h1\\">H1</h1>\\n<p>Content 1</p><h2 id=\\"h2\\">H2</h2>\\n<p>Content 2</p><h3 id=\\"h3\\">H3</h3>\\n<p>Content 3</p><h4 id=\\"h4\\">H4</h4>\\n<p>Content 4</p><h5 id=\\"h5\\">H5</h5>\\n<p>Content 5</p><h6 id=\\"h6\\">H6</h6>\\n<p>Content 6</p>";
export const Component = () => React.createElement("div", { dangerouslySetInnerHTML: { __html: htmlContent } });
export default Component;"
`;

exports[`vite-plugin-norg > React Mode > should generate correct component for basic norg file 1`] = `
"import React from "react";
export const metadata = {"author":"Drake Bott","tags":["\\"test\\", \\"basic\\""],"title":"Basic Norg"};
const htmlContent = "<h1 id=\\"main-title\\">Main Title</h1>\\n<p>Text with <i>bold</i>, <em>italic</em>, <code>code</code>, and <s>strikethrough</s>.</p><h2 id=\\"lists\\">Lists</h2>\\n<ul><li>Item 1</li></ul><ul><li>Item 2</li></ul><ol><li>Number 1</li></ol><ol><li>Number 2</li></ol><h2 id=\\"link\\">Link</h2>\\n<p><a href=\\"https://example.com\\" target=\\"_blank\\">link text</a></p>";
export const Component = () => React.createElement("div", { dangerouslySetInnerHTML: { __html: htmlContent } });
export default Component;"
`;

exports[`vite-plugin-norg > React Mode > should generate correct component for code blocks 1`] = `
"import React from "react";
export const metadata = {"title":"\\"Code Test\\""};
const htmlContent = "<h1 id=\\"code-blocks\\">Code Blocks</h1>\\n<h2 id=\\"javascript\\">JavaScript</h2>\\n<pre class=\\"language-javascript\\"><code class=\\"language-javascript\\">function hello() { return &quot;world&quot;; }\\n</code></pre><h2 id=\\"python\\">Python </h2>\\n<pre class=\\"language-python\\"><code class=\\"language-python\\">def add(a, b): return a + b\\n</code></pre><h2 id=\\"plain\\">Plain</h2>\\n<pre><code>plain text\\n</code></pre>";
export const Component = () => React.createElement("div", { dangerouslySetInnerHTML: { __html: htmlContent } });
export default Component;"
`;

exports[`vite-plugin-norg > React Mode > should generate correct component for minimal norg files 1`] = `
"import React from "react";
export const metadata = {};
const htmlContent = "<h1 id=\\"simple-title\\">Simple Title</h1>\\n<p>Basic content.</p>";
export const Component = () => React.createElement("div", { dangerouslySetInnerHTML: { __html: htmlContent } });
export default Component;"
`;

exports[`vite-plugin-norg > Svelte Mode > should generate correct component for all heading levels 1`] = `
"<script>
  const htmlContent = "<h1 id=\\"h1\\">H1</h1>\\n<p>Content 1</p><h2 id=\\"h2\\">H2</h2>\\n<p>Content 2</p><h3 id=\\"h3\\">H3</h3>\\n<p>Content 3</p><h4 id=\\"h4\\">H4</h4>\\n<p>Content 4</p><h5 id=\\"h5\\">H5</h5>\\n<p>Content 5</p><h6 id=\\"h6\\">H6</h6>\\n<p>Content 6</p>";
</script>
<script context="module">
  export const metadata = {"title":"\\"Headings\\""};
</script>
{@html htmlContent}"
`;

exports[`vite-plugin-norg > Svelte Mode > should generate correct component for basic norg file 1`] = `
"<script>
  const htmlContent = "<h1 id=\\"main-title\\">Main Title</h1>\\n<p>Text with <i>bold</i>, <em>italic</em>, <code>code</code>, and <s>strikethrough</s>.</p><h2 id=\\"lists\\">Lists</h2>\\n<ul><li>Item 1</li></ul><ul><li>Item 2</li></ul><ol><li>Number 1</li></ol><ol><li>Number 2</li></ol><h2 id=\\"link\\">Link</h2>\\n<p><a href=\\"https://example.com\\" target=\\"_blank\\">link text</a></p>";
</script>
<script context="module">
  export const metadata = {"author":"Drake Bott","tags":["\\"test\\", \\"basic\\""],"title":"Basic Norg"};
</script>
{@html htmlContent}"
`;

exports[`vite-plugin-norg > Svelte Mode > should generate correct component for code blocks 1`] = `
"<script>
  const htmlContent = "<h1 id=\\"code-blocks\\">Code Blocks</h1>\\n<h2 id=\\"javascript\\">JavaScript</h2>\\n<pre class=\\"language-javascript\\"><code class=\\"language-javascript\\">function hello() { return &quot;world&quot;; }\\n</code></pre><h2 id=\\"python\\">Python </h2>\\n<pre class=\\"language-python\\"><code class=\\"language-python\\">def add(a, b): return a + b\\n</code></pre><h2 id=\\"plain\\">Plain</h2>\\n<pre><code>plain text\\n</code></pre>";
</script>
<script context="module">
  export const metadata = {"title":"\\"Code Test\\""};
</script>
{@html htmlContent}"
`;

exports[`vite-plugin-norg > Svelte Mode > should generate correct component for minimal norg files 1`] = `
"<script>
  const htmlContent = "<h1 id=\\"simple-title\\">Simple Title</h1>\\n<p>Basic content.</p>";
</script>
<script context="module">
  export const metadata = {};
</script>
{@html htmlContent}"
`;
